Experimental Workflow interface

psfoley opened the pull request:
Closes https://github.com/intel/openfl/issues/627

This PR adds:

- A new workflow interface (inspired by the [Metaflow](https://github.com/Netflix/metaflow) project) - Define a chain of tasks to be performed on aggregator or collaborator. 
- Runtimes - A runtime defines the infrastructure where the workflow will be executed. 
    - [x] LocalRuntime - Run the workflow on a single node
        - Two backends supported: Execute a flow in a **single process** (default) or in multiple processes (for concurrent execution) with **[Ray](https://github.com/ray-project/ray)**
    - Additional runtimes supporting multi-node execution to be made available in a future release
- Partial compatibility with Metaflow Client for enhanced federated workflow debugging (note: all of these are optional)
    - [x] Save Data Artifacts
    - [x] Save stdout / stderr
    - [x] Generate Workflow DAG that can be loaded in HTML (only currently compatible in flows without loops)
- New Use Cases / Tutorials under the `openfl-tutorials/experimental` folder:
    - [x] Aggregator Validation after Collaborator Training
    - [x] Model Watermarking
    - [x] Privacy Meter - Measuring information leakage after model training
    - [x] Global Differential Privacy
    - [x] Vertical FL Examples
    - [x] Cyclic Institutional Incremental Training 


Remaining items before removing WIP tag:
- [x] Refactor logic in `LocalRuntime`
- [x] Merge PRs that ensure [no leakage between tasks](https://github.com/psfoley/openfl/pull/17)
- [x] Selective CI Pipelines for PRs labelled with `workflow_interface`
- [x] Comprehensive documentation
- [x] Review and update licenses

This PR includes contributions from:
@mansishr, @anindya-paul, @brandon-edwards, @changhongyan123, @scngupta-dsp, @ParthM-GitHub, @Greesh-N, @KeertiX, @igor-davidyuk, @aleksandr-mokrov, @itrushkin, @msheller, @hanboa

psfoley committed:
Remove CLA from OpenFL Repo

psfoley committed:
Merge branch 'intel:develop' into experimental-workflow-interface

psfoley committed:
Experimental workflow interface initial commit

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Added Vertical FL Example

psfoley committed:
WIP

psfoley committed:
LocalRuntime: Ray support with fallback to single process. Request GPUs through decorator. UI Code needs debugging for jupyter notebook (currently disabled). Bug fixes

psfoley committed:
merge commit

psfoley committed:
Fixed optimizer state in primary example

KeertiX committed:
Changes to support Flowgraph with Jupyter notebook

KeertiX committed:
Changes to support Flowgraph with Jupyter notebook

Greesh-N committed:
Test case for validating subset of collaborator works correctly.

KeertiX committed:
Test to validate references.

KeertiX committed:
renamed the file

psfoley committed:
Clean up next function

psfoley committed:
Refactored next function working as expected

psfoley committed:
Merge pull request #14 from scngupta-dsp/test_references

Test references

psfoley committed:
Merge pull request #13 from scngupta-dsp/testflow_subset_of_collab

Test case for validating subset of collaborator works correctly.

psfoley committed:
Merge branch 'experimental-workflow-interface' into dag-for-jupyternotebook

Greesh-N committed:
Added a mutex in MetaflowInterface.create_task

Greesh-N committed:
Test case for validating internalloops works correctly.

Greesh-N committed:
Test case for validating internalloops works correctly.

KeertiX committed:
Included graph code in metaflow_utils

KeertiX committed:
added pathspec to open graph on browser

psfoley committed:
Update workflow interface quickstart notebook

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

ParthM-GitHub committed:
Add files via upload

ParthM-GitHub committed:
Add files via upload

psfoley committed:
Cleaned up Workflow Interface 101 tutorial. Added tutorials 102 (aggregator validation), and 201 (requesting exclusive GPUs)

psfoley committed:
Updated with instructions for Google Colab. Cleared tutorial output

psfoley committed:
Minor fix to workflow interface GPU tutorial

ParthM-GitHub committed:
Fixed after one round filter attributes issue

Fix After one round filter attributes does not work properly

psfoley committed:
Documentation WIP

psfoley committed:
Merge pull request #16 from scngupta-dsp/testflow_internalloops

Test case for validating internal loops

Greesh-N committed:
corrected the spelling mistake

Commit:
Initial Version of Watermarking Tutorial

scngupta-dsp committed:
Delete Workflow_Interface_MNIST.ipynb

Unnecessary changes committed

psfoley committed:
Remove out of date example

psfoley committed:
Merge pull request #18 from scngupta-dsp/dag-for-jupyternotebook

Changes to support Flowgraph with Jupyter notebook

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Merge branch 'intel:develop' into experimental-workflow-interface

psfoley committed:
Remove old documentation from flow interface description. Section needs comprehensive descriptions still of interface / runtimes / features

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Add metaflow to 3rd party licenses

psfoley committed:
Adds CI tests for PRs tagged with 'workflow_interface'

psfoley committed:
CI Pipeline fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

ParthM-GitHub committed:
passing input attribute to checkpoint

Making changes to save attribute "input" in datastore.

ParthM-GitHub committed:
Removing commented code

Removing commented code

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
CI Fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merged experimental-workflow-interface into branch

psfoley committed:
Merge branch 'experimental-workflow-interface' into malformed-metaflow-datastore

psfoley committed:
Merge pull request #15 from scngupta-dsp/malformed-metaflow-datastore

Added a mutex in MetaflowInterface.create_task

psfoley committed:
Merging experimental-workflow-interface into branch

psfoley committed:
Fixed issues with test cases

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Remove unused dependencies

psfoley committed:
Revert changes to metaflow utils

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merge pull request #17 from scngupta-dsp/fixed-filter-attributes-issues-in-next

Fixed filter attributes issues in next

psfoley committed:
Fixing lint errors

psfoley committed:
Resolving lint suggestions

Signed-off-by: Patrick Foley <psfoley@gmail.com>

KeertiX committed:
updated the test case for testflow reference with include and exclude

Greesh-N committed:
Test Cases for Include and exclude works correctly.

anindya-paul committed:
Global DP Workflow

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

anindya-paul committed:
ML Privacy Meter Workflow

Co-authored-by: Hongyan Chang <hongyan.chang@intel.com>
Co-authored-by: Anindya S. Paul <anindya.s.paul@intel.com>
Co-authored-by: Brandon Edwards <brandon.edwards@intel.com>
Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

Commit:
Initial version of the Watermarking notebook in Tutorial format

anindya-paul committed:
GPU OOM fix.

Co-authored-by: Hongyan Chang <hongyan.chang@intel.com>
Co-authored-by: Anindya S. Paul <anindya.s.paul@intel.com>
Co-authored-by: Brandon Edwards <brandon.edwards@intel.com>
Co-authored-by: Patrick Foley <patrick.foley@intel.com>

anindya-paul committed:
Fix optimizer memory leak.

Co-authored-by: Anindya S. Paul <anindya.s.paul@intel.com>
Co-authored-by: Brandon Edwards <brandon.edwards@intel.com>
Co-authored-by: Mansi Sharma <mansi.sharma@intel.com>
Co-authored-by: Patrick Foley <patrick.foley@intel.com>
Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

anindya-paul committed:
Fix testing function to load tensors into device.

Co-authored-by: Hongyan Chang <hongyan.chang@intel.com>
Co-authored-by: Anindya S. Paul <anindya.s.paul@intel.com>
Co-authored-by: Brandon Edwards <brandon.edwards@intel.com>
Co-authored-by: Patrick Foley <patrick.foley@intel.com>
Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

Commit:
Update Summary 09/12/2022:
1. Generate Watermark dataset in notebook instead of downloading the entire WAFFLE archive
2. Installation of imagen from github source (required for Watermark dataset generation)
3. Updates to Tutorial text
4. Modularized Train function as it is being used in multiple places in the Workflow

Commit:
Reset and cleared outputs

psfoley committed:
Merge pull request #25 from scngupta-dsp/workflow-tutorial-watermarking

Workflow Tutorial for Watermarking

anindya-paul committed:
Delete dataset and dataloader attributes after each round.

Co-authored-by: Hongyan Chang <hongyan.chang@intel.com>
Co-authored-by: Anindya S. Paul <anindya.s.paul@intel.com>
Co-authored-by: Brandon Edwards <brandon.edwards@intel.com>
Co-authored-by: Patrick Foley patrick.foley@intel.com
Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

psfoley committed:
Resolve ray object store memory leak. Makes single process default option for LocalRuntime

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Merge pull request #24 from anindya-paul/Privacy-Meter

Privacy meter Workflow

psfoley committed:
Merge pull request #23 from anindya-paul/global-dp

Global DP Workflow

psfoley committed:
Merge pull request #20 from scngupta-dsp/testflow-references-include-exclude

Test case for validating references are not shared with collaborators when include and exclude are used.

psfoley committed:
Merge pull request #21 from scngupta-dsp/testflow-include-exclude

Test Cases for validating filtering of attributes

psfoley committed:
Update readme.md

Update image links

psfoley committed:
Pin requirements for workflow tests

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

scngupta-dsp committed:
Updated testflow_internalloop.py to fix errors

scngupta-dsp committed:
Updated to fix import errors and Lint issues

scngupta-dsp committed:
Added copyright to Test cases

psfoley committed:
Fix imports for tests

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merge pull request #26 from scngupta-dsp/experimental-workflow-interface

Fixed Lint Warnings & Import errors in Workflow Interface Tests

anindya-paul committed:
enable single_process or ray backend in local_runtime

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

psfoley committed:
Update flow_interface.rst

scngupta-dsp committed:
Fix lint warnings in openfl/experimental/placement/placement.py
- ./openfl/experimental/placement/placement.py:6:1: F401 'gc' imported but unused
- ./openfl/experimental/placement/placement.py:14:1: E302 expected 2 blank lines, found 1

scngupta-dsp committed:
Fix lint warnings in openfl/experimental/utilities/utils.py
-  ./openfl/experimental/utilities/utils.py:6:1: E302 expected 2 blank lines, found 1

scngupta-dsp committed:
Fix lint warnings in openfl/experimental/runtime/runtime.py
- ./openfl/experimental/runtime/runtime.py:24:100: E501 line too long (102 > 99 characters)
- ./openfl/experimental/runtime/runtime.py:35:26: E711 comparison to None should be 'if cond is not None:'
- ./openfl/experimental/runtime/runtime.py:60:16: F541 f-string is missing placeholders

scngupta-dsp committed:
Fix lint warnings in openfl/experimental/interface/fl_spec.py
- ./openfl/experimental/interface/fl_spec.py:19:1: F401 'threading.Lock' imported but unused
- ./openfl/experimental/interface/fl_spec.py:20:1: F401 'time.sleep' imported but unused
- ./openfl/experimental/interface/fl_spec.py:255:13: F841 local variable 'func_refs' is assigned to but never used
- ./openfl/experimental/interface/fl_spec.py:256:13: F841 local variable 'remote_functions' is assigned to but never used
- ./openfl/experimental/interface/fl_spec.py:257:13: F841 local variable 'remote_ctx_refs' is assigned to but never used

scngupta-dsp committed:
Fix lint warnings in openfl/experimental/utilities/metaflow_utils.py
- ./openfl/experimental/utilities/metaflow_utils.py:111:1: E800 Found commented out code
- ./openfl/experimental/utilities/metaflow_utils.py:128:24: C402 Unnecessary generator - rewrite as a dict comprehension.
- ./openfl/experimental/utilities/metaflow_utils.py:135:1: E800 Found commented out code
- ./openfl/experimental/utilities/metaflow_utils.py:152:1: E800 Found commented out code
- ./openfl/experimental/utilities/metaflow_utils.py:156:1: E800 Found commented out code
- ./openfl/experimental/utilities/metaflow_utils.py:260:61: FS001 '%' operator used
- ./openfl/experimental/utilities/metaflow_utils.py:264:21: F841 local variable 'e' is assigned to but never used
- ./openfl/experimental/utilities/metaflow_utils.py:276:60: FS001 '%' operator used
- ./openfl/experimental/utilities/metaflow_utils.py:280:25: F841 local variable 'e' is assigned to but never used
- ./openfl/experimental/utilities/metaflow_utils.py:282:21: F841 local variable 'e' is assigned to but never used
- ./openfl/experimental/utilities/metaflow_utils.py:394:42: FS002 '.format' used
- ./openfl/experimental/utilities/metaflow_utils.py:476:32: C408 Unnecessary dict call - rewrite as a literal.
- ./openfl/experimental/utilities/metaflow_utils.py:497:21: C408 Unnecessary dict call - rewrite as a literal.
- ./openfl/experimental/utilities/metaflow_utils.py:535:32: C408 Unnecessary dict call - rewrite as a literal.
- ./openfl/experimental/utilities/metaflow_utils.py:536:22: C408 Unnecessary dict call - rewrite as a literal.

mansishr commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,444 @@
+            self.next(self.check_round_completion)
+
+    # @collaborator                     # Uncomment this if you don't have GPU in the machine and want this application ro run on CPU instead 
+    @collaborator(num_gpus=1)           # Assuming GPU(s) is available in the machine
```
Somehow Ray remote is not being invoked, all processes are going to the first cuda device.

anindya-paul replied:
We set `backend='single_process'` by default as the Ray backend requires more system memory than the `single_process`, so we recommend ` backend='ray` on high-memory servers and systems.  The following will  demonstrate how to enable Ray backend through `LocalRuntime` if one wishes to experiment with Ray:
`local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='ray')`

psfoley replied:
@mansishr You're correct that in the case of the `single_process` backend for the `LocalRuntime`, all models will be sent to the same device. This should be okay because with the `single_process` backend, collaborators execute sequentially so only one GPU should be used at a time for this example. 

If there are multiple GPUs and sufficient memory in the system to take advantage of concurrent collaborator training, the Ray backend (enabled by setting `local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='ray')`) will automatically set  `CUDA_VISIBLE_DEVICES` within the task, and each collaborator will be able to use their own dedicated GPU without the user needing to make any change to the workflow definition. 

It's worth noting though that setting `num_gpus=1` is only honored by the Ray backend - the `single_process` backend will just ignore this request. I'll add a warning / info message when these placement decorator arguments are ignored for backends where they are not supported

mansishr commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,444 @@
+    collaborators = [Collaborator(name=name) for name in collaborator_names]
+    
+    if torch.cuda.is_available():
+        device = torch.device(f'cuda:0')  # This will enable Ray library to reserve available GPU(s) for the task
```
Devices should be maintained per collaborator. Right now, all processes are going to cuda 0.

anindya-paul committed:
fix lint errorrs with black reformatting.

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

anindya-paul committed:
fix minor typo

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

psfoley committed:
Update flow_interface.rst

psfoley committed:
Update flow_interface.rst

scngupta-dsp committed:
Merge branch 'psfoley:experimental-workflow-interface' into experimental-workflow-interface

ParthM-GitHub committed:
Refactored LocalRuntime and updated the runtime package file
	structure as follows:
1. Created seperate files for each Runtime
2. Refacoted excute_task method in LocalRuntime from FLSpec
3. Methods used commonly in both functions next, and execute_task
	are moved to utilities package runtime_utils.py file
4. Updated Watermarking Tutorial to correct the checkpoint
	usage (flflow.checkpoint to flflow._checkpoint)
5. Resolved flake8 suggestions in openfl/experimental package

ParthM-GitHub committed:
1. Updated Watermarking Tutorial to correct the checkpoint
	usage (flflow.checkpoint to flflow._checkpoint)

psfoley committed:
Add standardized requirements for workflow interface. Upgrade protobuf minor version

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Install common workflow requirements. Small updates to Privacy Meter

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Update global db requirements

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Vertical FL Examples

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Remove extra lines from notebook (no checkpointing for 2 party example)

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Merge branch 'experimental-workflow-interface' into experimental-workflow-interface

psfoley committed:
Merge pull request #28 from anindya-paul/experimental-workflow-interface

Fixed Lint Warnings for Privacy_Meter/ and Global_DP/ modules

psfoley committed:
Merge pull request #29 from scngupta-dsp/experimental-workflow-interface

Refactor logic in LocalRuntime

igor-davidyuk commented on openfl/experimental/runtime/local_runtime.py:
```diff
@@ -0,0 +1,177 @@
+    ):
+        """Use Local infrastructure to run the flow"""
+        super().__init__()
+        if backend not in ["ray", "single_process"]:
```
In my understanding, `ray` and `single_process` backends may be classified as separate runtimes. Does it make sense to have separate  `RayLocalRuntime` and `SPLocalRuntime` classes both implementing the `Runtime` interface?

psfoley replied:
@igor-davidyuk I'm glad you raised this. My intention was to create a hierarchy and logically divide `Runtime` to define where the workflow would run, and the `Backend` would specify how it would be executed. That way in the future there it will be clear if the user selects `LocalRuntime` the workflow will be executed locally, and `FederatedRuntime` runs the workflow across multiple nodes, and the user doesn't need to think about the backend unless they have good reason to do so. @mansishr, @aleksandr-mokrov, @itrushkin what are your thoughts about this?

igor-davidyuk replied:
Suppose we want to differentiate between runtimes and backends. In that case, we need a `backend` interface class to encapsulate the backend methods. 
`if else` approach is not sustainable if we would add more backends.

mansishr replied:
I like the idea of dividing `Runtime` to define where the flow would run, making it clear for the user. About `Backend` being an interface class, for now we only support `Ray` so using `if else` looks okay to me.

psfoley replied:
@igor-davidyuk and I discussed this offline. I agree a formal interface for the backend is more sustainable. I've added an issue for this to be addressed in the next release: https://github.com/intel/openfl/issues/676

igor-davidyuk commented on openfl/experimental/runtime/runtime.py:
```diff
@@ -0,0 +1,13 @@
+""" openfl.experimental.runtime module Runtime class."""
+
+
+class Runtime:
```
The specification for this component needs elaboration.

psfoley replied:
Added base class methods

ParthM-GitHub committed:
Fixed issue of aggregator privates are visible in collaborator steps via runtime.

psfoley committed:
Merge pull request #30 from scngupta-dsp/experimental-workflow-interface

Bug Fix: aggregator attributes should not be visible in collaborator steps

psfoley committed:
Update flow_interface.rst

scngupta-dsp committed:
Modified collaborator to be a private attribute of LocalRuntime so that collaborator private variables are not accessible to aggregators

scngupta-dsp committed:
Collaborator is made a private attribute of local runtime. Due to this local_runtime._collaborators is no longer accessible. Updated Workflow Interface testcases and Tutorials with this change

scngupta-dsp committed:
Added a new Test case to validate accessibility of private attributes
- Aggregator private attributes are accessible only in Aggregator steps
- Collaborator private attributes are accessible only in Collaborator steps

anindya-paul committed:
Addressing lint failures (WIP)

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

anindya-paul committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Fixed Lint issues

psfoley committed:
Merge pull request #33 from psfoley/lint-fixes

Lint fixes

psfoley committed:
Merge branch 'experimental-workflow-interface' into experimental-workflow-interface

psfoley committed:
Merge pull request #31 from scngupta-dsp/experimental-workflow-interface

[Bug]: Collaborator private attributes were accessible from Aggregator steps

psfoley committed:
Lint fix

psfoley committed:
Include more tests, validate both runtime backends

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Create Base Runtime Specification

Signed-off-by: Patrick Foley <psfoley@gmail.com>

anindya-paul committed:
Added back readme.md

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

anindya-paul committed:
Update readme.md

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

changhongyan123 committed:
add versions for packages

Signed-off-by: Hongyan chang <hongyan@comp.nus.edu.sg>

MasterSkepticista commented on docs/flow_interface.rst:
```diff
@@ -0,0 +1,287 @@
+    ExampleDedicatedGPUFlow(FLSpec):
+        ...
+        # We request one dedicated GPU for this task 
+        @collaborator(num_gpus=1)
```
What happens if `num_gpus` is set >1? Does the user have to write distributed dataparallel primitives? 
How are those devices enumerated to each process?
Alternatively, if the argument specifies `request_dedicated_gpu=True`, it would add clarity as to whether or not a user needs to write dataparallel primitives.

mansishr replied:
IMO, in that case, the workload is launched on a single GPU while other GPUs remain idle. According to [ray documentation](https://docs.ray.io/en/latest/ray-core/tasks/using-ray-with-gpus.html), calling `ray.remote()` internally calls `ray.get_gpu_ids()` which returns a list of available GPU IDs.

psfoley replied:
@MasterSkepticista `num_gpus=1` will likely be the most common usage, but there are other cases where using the functionality as implemented in ray will be useful. Setting  `num_gpus=0.5` allows two processes to run simultaneously on a single GPU. Users should be able to tune this parameter based on the size of their model / details of their workflow. One of the other possibilities is [an experiment that involves an ensemble of models](https://gitlab.com/frontierdevelopmentlab/astronaut-health/crisp), and by setting `num_gpus=2` the user could send different models to different GPUs. @mansishr mentioned one of the ways that device ids can be enumerated, but the other way is using the standard `os.environ["CUDA_VISIBLE_DEVICES"]`

anindya-paul replied:
@psfoley: following along the line what @MasterSkepticista asked: if a user enables PyTorch "[DISTRIBUTEDDATAPARALLEL](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html)" primitive for multi-GPU model training, will that cause conflict with Ray distributed orchestration?

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/README.md:
```diff
@@ -0,0 +1,18 @@
+
+## Global DP example
+
+Global DP implementation using the Opacus library (https://opacus.ai/) to perform global differentially private federated learning. At each round, a subset of collaborators are selected using a Poisson distribution over all collaborators, the selected collaborators perform local training with periodic clipping of their model delta (with respect to the current global model) to bound their contribu ...
```
Nitpick: The first sentence might need refinement, plus a line or two on what DP is, when does it help.

psfoley replied:
@mansishr could you address this?

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,654 @@
+# Fixing the seed for result repeatation: remove below to stop repeatable runs
+# ----------------------------------
+random_seed = 5495300300540669060
+torch.manual_seed(random_seed)
```
A more reliable way is to use local-scope RNGs. 
The current method fixes seed globally for the entire script and modules called thereafter. If certain modules change, they affect the state of all other modules later (even if repetitive states are guaranteed)
Details [here](https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator).

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,654 @@
+    return optim.SGD(model.parameters(), lr=learning_rate, momentum=momentum)
+
+
+def FedAvg(models, previous_global_model=None, dp_params=None):
```
General: Typehints and docstrings are helpful for end-users to write code without referring to the documentation.

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,654 @@
+        torch.cuda.empty_cache()
+        self.next(self.local_model_validation)
+
+    # Uncomment this if you don't have GPU in the machine
```
Another possibility where different flag naming helps.
`request_dedicated_gpu=True` vs `num_gpus=1` leaves the choice to the runtime. `CPU` is allocated if `GPU` is unavailable (with a warning since user requested one)

psfoley replied:
See the point above about keeping `num_gpus=` for consistency with Ray (and use cases where varying the number of requested GPUs is needed). I like the idea of fallback to CPU in the case where none are available - will add this as an issue to be addressed in the future

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/Workflow_Interface_Mnist_Implementation_1.py:
```diff
@@ -0,0 +1,654 @@
+            self.optimizer.load_state_dict(
+                deepcopy(self.optimizers[self.input].state_dict())
+            )
+            optimizer_to_device(optimizer=self.optimizer, device=self.device)
```
Would it be more efficient to push the params to device from `state_dict()`, before calling `self.optimizer.load_state_dict(...)`?

mansishr replied:
Won't make much of a difference for the MNIST example but @anindya-paul what do you think?

mansishr replied:
Fixed

anindya-paul replied:
It doesn't matter in this case as `self.optimizers` dictionary is initialized to CPU and hence state of `self.optimizers[self.input]` needs to move from CPU to GPU even if we load `self.optimizer` state to GPU first.

MasterSkepticista commented on openfl-tutorials/experimental/Global_DP/requirements_global_dp.txt:
```diff
@@ -0,0 +1,8 @@
+numpy==1.23.3
```
Strictly equal version requirements are primary causes of dependency loops. They work today, but not a few months down the line.
I suggest relaxing requirements for generic packages like `numpy`, `matplotlib` and others, wherever possible, to reduce conflicts when users install their own packages.

mansishr replied:
I agree, maybe we could change it to `numpy>=1.23.3` to use latest versions. @anindya-paul what do you think?

anindya-paul replied:
I have seen sometimes the latest version of NumPy causes issues. For example, the Privacy Meter example breaks with numpy==1.24.0 and hence we had to fix the Numpy version requirement there. In my experience, Numpy is one of the worst offenders in breaking interfaces and also sometimes causing confusing errors, so I'd suggest fixing the version that we have tested and knowing that it works with the release.

MasterSkepticista reviewed (commented):
Partial feedback on code reviewed so far:
- Please use typehints across functions, unless the function's behaviour is implicit or obvious. Typehints are a documentation in disguise. 
- Most functions have missing docstrings (by that, I mean an account of accepted values of an argument, default values etc., not just a literal expansion of what the function does)
- Given this new interface and growth of this repository, it would make sense to "consider" moving examples to a separate repository, and tag releases that are supposed to work for examples.

igor-davidyuk commented:
It is nice the PR comes with a vertical FL example.
I am wondering if it would be possible to implement cyclic training and cross-site validation schemes with FLfow interface.

mansishr commented on openfl/experimental/interface/fl_spec.py:
```diff
@@ -0,0 +1,178 @@
+            include_backup = deepcopy(self)
+            return_objs.append(include_backup)
+        return return_objs
+
```
Since we are not using `exclude_backup` and `include_backup` objects differently, could this be condensed to:

```
if "exclude" in kwargs or "include" in kwargs:
    return_objs = deepcopy(self)
return return_objs
```

psfoley replied:
Fixed

psfoley committed:
Addressing review comments

Commit:
fixing review comments

Signed-off-by: Mansi Sharma <mansi.sharma@intel.com>

psfoley committed:
Added type hints, docstrings. Distinguishing private methods from public

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley committed:
Fix circular dependencies resulting from type annotations

psfoley committed:
Add type annotations

psfoley committed:
Merge branch 'develop' into experimental-workflow-interface

psfoley committed:
Address lint naming errors

psfoley committed:
Merge pull request #37 from changhongyan123/privacy_meter_packages

Specify version for packages

psfoley committed:
More docstrings and type hints

psfoley committed:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

psfoley commented:
> Partial feedback on code reviewed so far:
> 
> * Please use typehints across functions, unless the function's behaviour is implicit or obvious. Typehints are a documentation in disguise.
> * Most functions have missing docstrings (by that, I mean an account of accepted values of an argument, default values etc., not just a literal expansion of what the function does)
> * Given this new interface and growth of this repository, it would make sense to "consider" moving examples to a separate repository, and tag releases that are supposed to work for examples.

@MasterSkepticista thanks for the feedback! I've added type hints and expanded docstrings for user facing functions. @anindya-paul is working on some of the other comments specific to Global DP and Privacy Meter.

psfoley committed:
flake8 fixes

Signed-off-by: Patrick Foley <psfoley@gmail.com>

psfoley committed:
Added cyclic institutional incremental learning example

psfoley commented:
> It is nice the PR comes with a vertical FL example. I am wondering if it would be possible to implement cyclic training and cross-site validation schemes with FL Flow interface.

@igor-davidyuk Excellent idea. I've added a cyclic training example notebook to this PR here that compares that approach with standard horizontal FL under various IID / Non-IID scenarios: https://github.com/intel/openfl/blob/554f674b71b7bc5346e65a98752d4959210492eb/openfl-tutorials/experimental/Workflow_Interface_103_Cyclic_Institutional_Incremental_Learning.ipynb

mansishr commented on openfl/experimental/placement/placement.py:
```diff
@@ -0,0 +1,151 @@
+        self.remote_contexts = []
+
+    def ray_call_put(self, ctx, func):
+        remote_to_exec = make_remote(func, num_gpus=func.num_gpus)
```
Docstrings can be added here

mansishr commented on openfl/experimental/placement/placement.py:
```diff
@@ -0,0 +1,151 @@
+        del ref_ctx
+
+    def get_remote_clones(self):
+        clones = deepcopy(ray.get(self.remote_functions))
```
add docstrings

mansishr commented on tests/github/experimental/testflow_reference.py:
```diff
@@ -0,0 +1,369 @@
+
+
+def validate_agg_attr_ref(agg_attrs, agg_obj):
+    attr_flag = False
```
add docstring.

psfoley replied:
Added

mansishr commented on tests/github/experimental/testflow_reference_with_exclude.py:
```diff
@@ -0,0 +1,335 @@
+
+
+def find_match_ref_at_step(collab_attr_list, all_collborators):
+    collab_names = all_collborators.keys()
```
add docstring.

psfoley replied:
Added

psfoley committed:
Resolve review comments - add docstrings to tests

Commit:
using Generator class for fixing seed

Signed-off-by: Mansi Sharma <mansi.sharma@intel.com>

Commit:
Merge branch 'experimental-workflow-interface' of https://github.com/psfoley/openfl into experimental-workflow-interface

mansishr reviewed (approved):
LGTM, great work!

psfoley committed:
Make flake8 exceptions

anindya-paul committed:
added doscstrings to several functions for better user understanding about the functionality

Signed-off-by: Paul, Anindya S <anindya.s.paul@intel.com>

psfoley committed:
Merge pull request #38 from anindya-paul/experimental-workflow-interface

Added docstrings to several functions for better user understanding

psfoley committed:
Fix jupyter notebook links. Update documentation / create links from README

psfoley committed:
More README updates
